UrlEncode user input in client

vdemeester opened the pull request:
This is linked to #29126, this disallow any `../` on a query from the client. This could lead to some really messed up queries.

Currently, a request with `../something` like `docker network rm ../foo/bar` returns a 301 which is not really handled at all. 

```bash
$ docker network rm ../foo/bar
# […]
resp &{301 Moved Permanently 301 HTTP/1.1 1 1 map[Content-Type:[text/plain; charset=utf-8] Location:[/toto] Date:[Mon, 05 Dec 2016 11:16:35 GMT] Content-Length:[0]] 0xc42000e500 0 [] false false map[] 0xc420424000 <nil>}
../foo/bar
```

We should discuss what to do about `301` http code too.

/cc @thaJeztah @dnephin @justincormack @cpuguy83

🐸 

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

thaJeztah commented:
Looks like build is failing;

```
13:47:19 # github.com/docker/docker/client
13:47:19 client/client_test.go:161: multiple-value c.getAPIPath() in single-value context
13:47:19 FAIL	github.com/docker/docker/client [build failed]
```

vdemeester commented:
@thaJeztah oups :sweat:

duglin commented on client/client.go:
```diff
@@ -191,8 +191,11 @@ func (cli *Client) Close() error {
-func (cli *Client) getAPIPath(p string, query url.Values) string {
+func (cli *Client) getAPIPath(p string, query url.Values) (string, error) {
 	var apiPath string
+	if strings.Contains(p, "../") {
```
I'm not sure this is correct since this would also ban `foo../` which isn't what we want.
I know we may not ever have "foo.." as a valid part of the URL but I think that's a different topic. Perhaps changing it to "/../" if we really want to do this.

thaJeztah replied:
I was discussing earlier with @vdemeester if there are other potentially dangerous variations. If we could detect something like `if value !== escapedValue(value)`

vdemeester replied:
good point :smile:

duglin commented:
I would prefer a different approach:
1 - the client should not try to do URL tweaking - it should take all user names "as is" and just URL escape them so that the HTTP layer doesn't see it as ".." at all.  It should see it as "%2E%2E" instead.
2 - the daemon should be the one to block the use of ".." in an incoming message since people can do just "curl" cmds to by-pass the docker cli.  This assumes that the use of ".." is seen by our code at all - which it may not be - the http layer may send back a 301.  In which case I think we should either do the redirect or show an error - instead of being silent like we are today.

vdemeester commented:
I do agree with `1.` :angel:

thaJeztah commented:
>  the daemon should be the one to block the use of ".." in an incoming message

yeah, I don't think the daemon will see those; `/v1.25/networks/../containers/foo` is the same as `/v1.25/containers/foo`. That's not a problem directly if people are calling the API directly, but allowing it from the CLI is odd, and something we (I think) should prevent.

Of course, unless an authorization plugin  is installed, having access to the API allows you to do any action, so we're not protecting against anything, but it would be a good thing to disallow it.

vdemeester commented:
Updated the PR to encode the path instead of checking for `../` :angel: (the returned error is a little bit odd though >_<)

```bash
$ docker network rm ../foo/bar
Error response from daemon: page not found
```

duglin commented:
That's odd, I would expect the same error as this:
```
docker network rm asdasd
Error response from daemon: network asdasd not found
```
if we're not seeing that then I suspect its because the escaping isn't being done right, on one or both sides because `../foo/bar` should just be a name/string, not part of the path.

vdemeester commented on client/client.go:
```diff
@@ -193,6 +193,7 @@ func (cli *Client) Close() error {
 // It appends the query parameters to the path if they are not empty.
 func (cli *Client) getAPIPath(p string, query url.Values) string {
 	var apiPath string
+	p = url.QueryEscape(p)
```
Hum right… :sweat: wished I didn't have to do that but… yeah

duglin commented:
See: https://github.com/gorilla/mux/issues/132
and https://github.com/docker/docker/pull/16577
just remembered

vdemeester commented:
```
# docker network rm ../volumes/toto
Error response from daemon: network ..%2Fvolumes%2Ftoto not found
```

vdemeester commented:
@duglin just did the `remove` endpoint but we should do that for all :sweat:

duglin commented:
yup - easy way to bump up your commit count  LOL

vdemeester commented:
I'll do all at once :angel: so just 1 commit :stuck_out_tongue_closed_eyes:

vdemeester committed:
Encode path in client

To make sure we %dDisallow ../ in client http request path for example.
This could lead to some really messed up queries.

Signed-off-by: Vincent Demeester <vincent@sbr.pm>

thaJeztah commented:
sgtm

vdemeester commented:
Hum CI failures are not the one I though… I'll have to look into it 😅

thaJeztah commented:
Removing this from the 1.13 milestone, as this is too risky to include now

vdemeester commented:
Sounds good to me 😇

vdemeester commented:
Hum, I'll close it for now as I don't think this is the right fix… I'll take another look.
